plonky2/examples/square_root.rs

```diff
@@ -0,0 +1,35 @@
+use anyhow::Result;
+use plonky2::field::types::Field;
+use plonky2::iop::witness::PartialWitness;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use plonky2::plonk::circuit_data::CircuitConfig;
+use plonky2::plonk::config::{GenericConfig, PoseidonGoldilocksConfig};
+
+fn main() -> Result<()> {
+    const D: usize = 2;
+    type C = PoseidonGoldilocksConfig;
+    type F = <C as GenericConfig<D>>::F;
+
+    let config = CircuitConfig::standard_recursion_config();
+
+    let pw = PartialWitness::new();
+    let mut builder = CircuitBuilder::<F, D>::new(config);
+
+    let x = F::rand();
+    let x_squared = x * x;
+    let x_target = builder.constant(x);
+    let x_squared_target = builder.constant(x_squared);
+
+    let x_squared_computed = builder.mul(x_target, x_target);
+    builder.connect(x_squared_target, x_squared_computed);
```

Comment:
Ah okay! Does it look right now? 🤔

Reply:
Oh oops — I realized that in my code, the generator isn't actually necessary, because I'm just generating multiplication...

Reply:
Were you thinking of having a generator to find the square root _from_ the square? Using Tonelli-Shanks or something?

Reply:
> Were you thinking of having a generator to find the square root _from_ the square? Using Tonelli-Shanks or something?

Yeah that's what I was thinking. But I forgot that computing the square root in the Goldilocks field is a bit involved (since `p=4k+1`). So two options would be
- Implement a square root algorithm. We had one in plonky1 that should still work https://github.com/mir-protocol/plonky/blob/3403436d3df4bfb2b530e2ca2a78c9c2a9b9a8f0/src/field/field.rs#L440.
- Change this example to something like computing the inverse.  This would be somewhat equivalent to the square-root example ("I know a `y` st `y*x=1`" instead of "I know a `y` st `y*y=x`"), and writing the generator would be easier since inversion is already implemented.